crates/anemo/src/network/wire.rs

```diff
@@ -29,6 +29,33 @@ pub(crate) fn network_message_frame_codec() -> LengthDelimitedCodec {
         .new_codec()
 }
 
+/// Anemo requires mTLS in order to ensure that both sides of the connections are authenticated by
+/// the other. This is specifically required so that regaurdless of which peer initiaties a
+/// connection, both sides will be able to know the PeerId of the other side. One challenge with
+/// this is that due to the ordering of how certs are exchanged, the client side may think the
+/// connection is fully established when in reality the server may still reject the connection. To
```

Comment:
Seems like an extra round trip shouldn't be necessary at least as far as the QUIC protocol is concerned. Is this becuase quinn lacks a "hook" for anemo code to inspect the client identity as part of QUIC establishment, and only lets you look at who the client is after the QUIC handshake is finished? Or what's the missing thing?

Reply:
So this isn't a quinn issue but actually sort of an issue with the QUIC protocol (at least my read of it is) take a look at the commit message for the commit that introduces this functionality:

```
wire: introduce a lightweight handshake
Introduce a lightweight handshake that is performed after the server
side of a connection is able to fully complete the mTLS process and
verify the client side's certificate. Adding this handshake eliminates
the client side from prematurely thinking that the connection has been
fully established.

Before this patch the process looked like:

1. Client initiates contact with the server
2. Server presents its TLS certificate to the client
3. Client verifies the server's certificate <-- Client thinks the connection is complete
4. Client presents its TLS certificate to the server
5. Server verifies the client's certificate <-- Server thinks the connection is complete

After this patch this looks like:

1. Client initiates contact with the server
2. Server presents its TLS certificate to the client
3. Client verifies the server's certificate
4. Client presents its TLS certificate to the server
5. Server verifies the client's certificate
6. Server sends ACK <-- Server thinks the connection is complete
7. Client recieves ACK <-- Client thinks the connection is complete
```